### PR TITLE
Refactor louvain 

### LINKED ISSUE #2493: MNMG Louvain - Python
The python implementation of Louvain still uses the legacy implementation.  There is already an SG and MG Louvain available in PLC (pylibcugraph).

Existing tests should be sufficient, it should just be a matter of refactoring the python code.

### PR DESCRIPTION
**Pull request by jnke2016**
This PR refactors Louvain by leveraging the C API and does minor updates to the python tests
closes #2493

### CODE REVIEW

File: python/pylibcugraph/pylibcugraph/louvain.pyx [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+    get_c_type_from_numpy_type,
+    create_cugraph_type_erased_device_array_view_from_py_obj,
+)
+
+
+def louvain(ResourceHandle resource_handle,
+                          _GPUGraph graph,
+                          size_t max_level,
+                          double resolution,
+                          bool_t do_expensive_check):
```
Comment (rlratzel) [inline python/pylibcugraph/pylibcugraph/louvain.pyx:62]: indentation

File: python/pylibcugraph/pylibcugraph/louvain.pyx [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+
+    Returns
+    -------
+    # FIXME: no modularity_score is returned liek in the legacy implementation
+    A tuple of device arrays, conataining the heirarchical clustering vertices
+    and clusters.
+
+    Examples
+    --------
+    # FIXME: No example yet
```
Comment (rlratzel) [inline python/pylibcugraph/pylibcugraph/louvain.pyx:100]: Can these FIXMEs be addressed (at least the return value including the modularity score)?

File: python/pylibcugraph/pylibcugraph/louvain.pyx [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    try:
+        import cupy
+    except ModuleNotFoundError:
+        raise RuntimeError("louvain requires the cupy package, which could "
+                           "not be imported")
+    try:
+        import numpy
+    except ModuleNotFoundError:
+        raise RuntimeError("louvain requires the numpy package, which could "
+                           "not be imported")
```
Comment (rlratzel) [inline python/pylibcugraph/pylibcugraph/louvain.pyx:115]: Can the FIXME and the imports here can be removed?

File: python/pylibcugraph/pylibcugraph/louvain.pyx [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    cdef cugraph_type_erased_device_array_view_t* vertices_ptr = \
+        cugraph_heirarchical_clustering_result_get_vertices(result_ptr)
+    cdef cugraph_type_erased_device_array_view_t* clusters_ptr = \
+        cugraph_heirarchical_clustering_result_get_clusters(result_ptr)
+    cdef double modularity = \
+        cugraph_heirarchical_clustering_result_get_modularity(result_ptr)
+
+    cupy_vertices = copy_to_cupy_array(c_resource_handle_ptr, vertices_ptr)
+    cupy_clusters = copy_to_cupy_array(c_resource_handle_ptr, clusters_ptr)
+    #cupy_modularity = copy_to_cupy_array(c_resource_handle_ptr, modularity_ptr)
```
Comment (rlratzel) [inline python/pylibcugraph/pylibcugraph/louvain.pyx:144]: Can this be removed?

**Review by rlratzel — CHANGES_REQUESTED** [review]
Just a few minor things

File: python/pylibcugraph/pylibcugraph/louvain.pyx (hunk 3 above)
Comment (jnke2016) [reply python/pylibcugraph/pylibcugraph/louvain.pyx:115]: Should I go for a `hard dependencies` or `import_optional()`

Comment (rlratzel) [reply python/pylibcugraph/pylibcugraph/louvain.pyx]: Neither, I think you can just remove them since they're no longer needed.

**Review by rlratzel — APPROVED** [review]

**Comment by BradReesWork** [comment]
rerun tests

**Review by ChuckHastings — APPROVED** [review]
The C/C++ changes here were part of a different PR.  I suspect that merging the branch would have removed them from this review.

In any case, they look the same to me.

**Comment by jnke2016** [comment]
rerun tests

**Comment by ChuckHastings** [comment]
rerun tests

**Comment by BradReesWork** [comment]
rerun tests

**Comment by BradReesWork** [comment]
@gpucibot merge

**Pull request merged**